### PR TITLE
Added CLI command to update noobaa account for default resource

### PR DESCRIPTION
**Pull request by achouhan09**
Signed-off-by: Aayush Chouhan <achouhan@redhat.com>

### Explain the changes
1.  Implemented a CLI command to update the default resource in noobaa account, that helps us in deletion of noobaa default backingstore.

### CODE REVIEW

File: pkg/noobaaaccount/noobaaaccount.go [hunk 1]
```diff
@@ -224,6 +236,85 @@ func RunCreate(cmd *cobra.Command, args []string) {
...
+
+	if isResourceBackingStore && isResourceNamespaceStore {
+		log.Fatalf(`❌  got BackingStore and NamespaceStore %q in namespace %q`,
+			newDefaultResource, options.Namespace)
+	} else if !isResourceBackingStore && !isResourceNamespaceStore {
+		log.Fatalf(`❌ Could not get BackingStore or NamespaceStore %q in namespace %q`,
+			newDefaultResource, options.Namespace)
+	}
+
+	if !util.KubeCheck(noobaaAccount) && (name != "admin@noobaa.io") {
```
Comment (liranmauda) [inline pkg/noobaaaccount/noobaaaccount.go:270]: The account `admin@noobaa.io` doesn't have CRD, it is internal.
having this condition means that the `admin@noobaa.io` is not updatable. 
```suggestion
	if !util.KubeCheck(noobaaAccount)  {
```

**Review by liranmauda — CHANGES_REQUESTED** [review]

File: pkg/noobaaaccount/noobaaaccount.go [hunk 2]
```diff
@@ -224,6 +236,85 @@ func RunCreate(cmd *cobra.Command, args []string) {
...
+// RunUpdate runs a CLI command
+func RunUpdate(cmd *cobra.Command, args []string) {
+	log := util.Logger()
+
+	if len(args) != 1 || args[0] == "" {
+		log.Fatalf(`❌ Missing expected arguments: <noobaa-account-name> %s`, cmd.UsageString())
+	}
+	name := args[0]
+
+	newDefaultResource, _ := cmd.Flags().GetString("new_default_resource")
```
[the remark this answers is not in the document]
Comment (achouhan09) [reply pkg/noobaaaccount/noobaaaccount.go:248]: Sure, will do that.

File: pkg/noobaaaccount/noobaaaccount.go [hunk 3]
```diff
@@ -224,6 +236,85 @@ func RunCreate(cmd *cobra.Command, args []string) {
...
+
+	o := util.KubeObject(bundle.File_deploy_crds_noobaa_io_v1alpha1_noobaaaccount_cr_yaml)
+	noobaaAccount := o.(*nbv1.NooBaaAccount)
+	noobaaAccount.Name = name
+	noobaaAccount.Namespace = options.Namespace
+
+	if newDefaultResource == "" {
+		log.Fatalf(`❌ New Default Resource must be provided as flag to update`)
+	}
```
[the remark this answers is not in the document]
Comment (achouhan09) [reply pkg/noobaaaccount/noobaaaccount.go:257]: Sure, Thanks.

File: pkg/noobaaaccount/noobaaaccount.go (hunk 1 above)
Comment (achouhan09) [reply pkg/noobaaaccount/noobaaaccount.go:270]: Okay.

File: test/cli/test_cli_functions.sh [hunk 4]
```diff
@@ -634,9 +634,61 @@ function account_cycle {
...
+    account_update ""
+    # update crd account default resource
+    account_update "crd"
     echo_time "✅  noobaa account cycle is done"
 }
 
+function account_update {
+    local crd=${1}
+    echo_time "💬  Checking for ${crd} account updation"
+    local res="backing-store-upd"
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:647]: This variable is meant for new_default_resource and not res. 
we can change the name to `new_default_resource` and change all the calls for it accordingly
```suggestion
    local new_default_resource="backing-store-upd"
```

File: test/cli/test_cli_functions.sh [hunk 5]
```diff
@@ -634,9 +634,61 @@ function account_cycle {
...
+    if [ "$(kuberun get backingstore | grep -w ${res} | wc -l)" != "1" ]
+    then
+        echo_time "💬  Creating backingstore ${res}"
+        test_noobaa backingstore create pv-pool ${res} --num-volumes 1 --pv-size-gb 16
+    fi
+    if [ "${crd}" == "" ]
+    then
+        local name=$(test_noobaa api account list_accounts {} -o json | jq -r '.accounts[0].email')
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:655]:
```suggestion
        local account_name=$(test_noobaa api account list_accounts {} -o json | jq -r '.accounts[0].email')
```

File: test/cli/test_cli_functions.sh [hunk 6]
```diff
@@ -634,9 +634,61 @@ function account_cycle {
...
+        local prev=$(test_noobaa api account list_accounts {} -o json | jq -r '.accounts[0].default_resource')
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:656]:
```suggestion
        local default_resource=$(test_noobaa api account list_accounts {} -o json | jq -r '.accounts[0].default_resource')
```

File: test/cli/test_cli_functions.sh [hunk 7]
```diff
@@ -634,9 +634,61 @@ function account_cycle {
...
+        echo_time "💬  Updating ${res} as default_resource for noobaa account"
+        test_noobaa account update ${name} --new_default_resource=${res}
+        local curr=$(test_noobaa api account list_accounts {} -o json | jq -r '.accounts[0].default_resource')
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:659]:
```suggestion
        local curr_default_resource=$(test_noobaa api account list_accounts {} -o json | jq -r '.accounts[0].default_resource')
```

File: test/cli/test_cli_functions.sh [hunk 8]
```diff
@@ -634,9 +634,61 @@ function account_cycle {
...
+        if [ "${prev}" == "${curr}" ]
+        then
+            echo_time "❌ Looks like account not updated, Exiting"
+            exit 1
+        else
+            echo_time "✅  Update account default_resource successful"
+            test_noobaa account update ${name} --new_default_resource=${prev}
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:666]:
```suggestion
            echo_time "✅  Update account default_resource successful"
            echo_time "💬  Updating ${prev} as default_resource for noobaa account"
            test_noobaa account update ${name} --new_default_resource=${prev}
```
We should also also verify that it worked.

File: test/cli/test_cli_functions.sh [hunk 9]
```diff
@@ -634,9 +634,61 @@ function account_cycle {
...
+    # update account default resource
+    account_update ""
+    # update crd account default resource
+    account_update "crd"
     echo_time "✅  noobaa account cycle is done"
 }
 
+function account_update {
+    local crd=${1}
+    echo_time "💬  Checking for ${crd} account updation"
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:646]:
```suggestion
    if [ -z ${crd} ]
    then
          echo_time "💬  Checking account update cycle"
    else
         echo_time "💬  Checking crd based account update cycle"
    fi
```

File: test/cli/test_cli_functions.sh [hunk 10]
```diff
@@ -634,9 +634,61 @@ function account_cycle {
     account_reset_password "admin@noobaa.io"
     # testing nsfs accounts
     # account_nsfs_cycle TODO re-enable.
+    # update account default resource
+    account_update ""
+    # update crd account default resource
+    account_update "crd"
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:640]: we just want to check if crd variable is not empty so here we can leave it:
```suggestion
    # update account default resource
    account_update 
    # update crd account default resource
    account_update "crd"
```

File: test/cli/test_cli_functions.sh [hunk 11]
```diff
@@ -634,9 +634,61 @@ function account_cycle {
...
+function account_update {
+    local crd=${1}
+    echo_time "💬  Checking for ${crd} account updation"
+    local res="backing-store-upd"
+    if [ "$(kuberun get backingstore | grep -w ${res} | wc -l)" != "1" ]
+    then
+        echo_time "💬  Creating backingstore ${res}"
+        test_noobaa backingstore create pv-pool ${res} --num-volumes 1 --pv-size-gb 16
+    fi
+    if [ "${crd}" == "" ]
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:653]:
```suggestion
    if [ -z ${crd} ]
```

File: test/cli/test_cli_functions.sh [hunk 12]
```diff
@@ -634,9 +634,61 @@ function account_cycle {
...
+        then
+            echo_time "❌ Looks like account not updated, Exiting"
+            exit 1
+        else
+            echo_time "✅  Update account default_resource successful"
+            test_noobaa account update ${name} --new_default_resource=${prev}
+        fi
+    fi
+    if [ "${crd}" == "crd" ]
+    then
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:670]:
```suggestion
    else
```

File: test/cli/test_cli_functions.sh [hunk 13]
```diff
@@ -634,9 +634,61 @@ function account_cycle {
...
+        if [ "${prev}" == "${curr}" ]
+        then
+            echo_time "❌ Looks like crd account not updated, Exiting"
+            exit 1
+        else
+            echo_time "✅  Update crd account default_resource successful"
+            echo_time "💬  Deleting crd account ${name}"
+            test_noobaa account delete ${name}
+            echo_time "💬  Deleting backingstore ${res}"
+            test_noobaa backingstore delete ${res}
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:687]: We should do this outside of this function, it is not related to the account update flow, but to the `account_cycle`

**Review by liranmauda — CHANGES_REQUESTED** [review]

File: pkg/noobaaaccount/noobaaaccount.go [hunk 14]
```diff
@@ -224,6 +236,77 @@ func RunCreate(cmd *cobra.Command, args []string) {
...
+		if WaitReady(noobaaAccount) {
+			log.Printf("")
+			log.Printf("")
+			RunStatus(cmd, args)
+		}
+	} else {
+		updateAccountS3AccessParams := nb.UpdateAccountS3AccessParams{
+			Email:           name,
+			DefaultResource: &newDefaultResource,
+			S3Access:        true,
```
Comment (liranmauda) [inline pkg/noobaaaccount/noobaaaccount.go:294]: We should take the `S3Access` from the account and keep it's current value

File: test/cli/test_cli_functions.sh (hunk 8 above)
Comment (achouhan09) [reply test/cli/test_cli_functions.sh]: Added check.

File: test/cli/test_cli_functions.sh (hunk 13 above)
Comment (achouhan09) [reply test/cli/test_cli_functions.sh]: Actually this new backingstore and crd based account is created for this specific case of `account_update` check. If we want to put them outside the function then we need to pass there name as param to `account_update` and also create that new backingstore and crd account inside the `account_cycle` check, and we can delete them after the cycle is complete in `account_cycle`.

File: test/cli/test_cli_functions.sh [hunk 15]
```diff
@@ -634,9 +634,66 @@ function account_cycle {
...
+    fi
+    if [ -z ${crd} ]
+    then
+        echo_time "💬  Checking account update cycle"
+        local account_name=$(test_noobaa api account list_accounts {} -o json | jq -r '.accounts[0].email')
+        local default_resource=$(test_noobaa api account list_accounts {} -o json | jq -r '.accounts[0].default_resource')
+        echo_time "💬  Updating ${new_default_resource} as default_resource for noobaa account"
+        test_noobaa account update ${account_name} --new_default_resource=${new_default_resource}
+        local curr_default_resource=$(test_noobaa api account list_accounts {} -o json | jq -r '.accounts[0].default_resource')
+        if [ "${default_resource}" == "${curr_default_resource}" ]
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:660]:
```suggestion
        if [ "${new_default_resource}" != "${curr_default_resource}" ]
```

File: test/cli/test_cli_functions.sh [hunk 16]
```diff
@@ -634,9 +634,66 @@ function account_cycle {
...
+    else
+        echo_time "💬  Checking crd based account update cycle"
+        local account_name="test-account"
+        echo_time "💬  Creating crd account ${account_name}"
+        test_noobaa account create ${account_name}
+        local default_resource=$(kuberun get NoobaaAccount ${account_name} -n test -o json | jq -r '.spec.default_resource')
+        echo_time "💬  Updating ${new_default_resource} as default_resource for noobaa account"
+        test_noobaa account update ${account_name} --new_default_resource=${new_default_resource}
+        local curr_default_resource=$(kuberun get NoobaaAccount ${account_name} -n test -o json | jq -r '.spec.default_resource')
+        if [ "${default_resource}" == "${curr_default_resource}" ]
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:683]:
```suggestion
        if [ "${new_default_resource}" != "${curr_default_resource}" ]
```

**Review by liranmauda — APPROVED** [review]
LGTM

**Comment by achouhan09** [comment]
@dannyzaken @naveenpaul1 Please have a look when you get time. Thanks!

**Pull request merged**